Libplanet.Net/Consensus/Context.Mutate.cs

```diff
@@ -0,0 +1,361 @@
+        /// </exception>
+        /// <exception cref="InvalidValidatorVoteMessageException">Thrown when the signature of
+        /// <see cref="Vote"/> is invalid or the <see cref="Vote"/> is not signed by any validator
+        /// of this context.
+        /// </exception>
+        internal void AddMessage(ConsensusMessage message)
+        {
+            if (message.Height != Height)
+            {
+                throw new InvalidHeightMessageException(
+                    "Height of message differs with working height.  " +
+                    $"(expected: {Height}, actual: {message.Height})",
+                    message);
+            }
+
+            if (message is ConsensusPropose propose)
+            {
+                if (!propose.Validator.Equals(Proposer(message.Round)))
+                {
+                    throw new InvalidProposerProposeMessageException(
+                        "Proposer for the height " +
+                        $"{message.Height} and round {message.Round} is invalid.  " +
+                        $"(expected: Height: {message.Height}, Round: {message.Round}, " +
+                        $"Proposer: {message.Validator} / " +
+                        $"actual: Height: {Height}, Round: {Round}, " +
+                        $"Proposer: {Proposer(message.Round)})",
+                        message);
+                }
+
+                if (message.BlockHash.Equals(default(BlockHash)))
+                {
+                    throw new InvalidBlockProposeMessageException(
+                        "Cannot propose a null block.",
+                        message);
+                }
+            }
+
+            if (message is ConsensusVote vote &&
+                (!vote.Validator.Equals(vote.ProposeVote.Validator) ||
+                 !vote.ProposeVote.Verify(vote.Validator) ||
+                 !_validators.Contains(vote.Validator)))
+            {
+                throw new InvalidValidatorVoteMessageException(
+                    "Received ConsensusVote message is made by invalid validator.",
+                    vote);
+            }
+
+            if (message is ConsensusCommit commit &&
+                (!commit.Validator.Equals(commit.CommitVote.Validator) ||
+                 !commit.CommitVote.Verify(commit.Validator) ||
+                 !_validators.Contains(commit.Validator)))
+            {
+                throw new InvalidValidatorVoteMessageException(
+                    "Received ConsensusCommit message is made by invalid validator.",
+                    commit);
+            }
+
+            lock (_messagesInRoundLock)
+            {
+                if (!_messagesInRound.ContainsKey(message.Round))
+                {
+                    _messagesInRound.TryAdd(message.Round, new HashSet<ConsensusMessage>());
+                }
+
+                // TODO: Prevent duplicated messages adding.
+                _messagesInRound[message.Round].Add(message);
```

Comment:
Is `HashSet` added for preventing duplicated message?

Reply:
`HashSet` is not added. It was there already. We might have to look into this further in another issue. 😗